### PR TITLE
KBC-2586 trigger can be manipulated by token with access to component

### PR DESCRIPTION
**Pull request by fejjo**
Jira: KBC-2586
MainPR: https://github.com/keboola/connection/pull/3700
Before asking for review make sure that:

- [x] New client method(s) has tests
- [x] Apiary file is updated
- [x] You declared if there is a BC break or not (will affect next release of a client)

---

### CODE REVIEW

**Comment by tomasfejfar** [comment]
👀

**Review by tomasfejfar — APPROVED** [review]
Upravil bych jen trochu wording a upravil jsem ve fixupu jednu proměnnou

File: apiary.apib [hunk 1]
```diff
@@ -8187,7 +8187,7 @@ Gets stats for a given Run Id. Provides the count of imported/exported tables an
 ###Create [POST /v2/storage/triggers/]
 Created a trigger listening on set of tables
 
-Token has to be [*master*](https://help.keboola.com/storage/tokens/) (then `runWithTokenId` can be also master). If it isn't master, it has to have `canManageBuckets` permission or `componentAccess` permission for component `keboola.orchestrator`. Applies also on Update and Delete operation
+Token has to be [*master*](https://help.keboola.com/storage/tokens/) (then `runWithTokenId` can be also master). If it isn't master, it has to have `canManageBuckets` permission or `componentAccess` permission for component which you desire create a trigger. Applies also on Update and Delete operation
```
[the remark this answers is not in the document]
Comment (fejjo) [reply apiary.apib]: Dik upravene

File: tests/Common/TriggersTest.php [hunk 2]
```diff
@@ -298,6 +298,124 @@ public function testUpdateTriggerCreatedByMasterToken(TokenCreateOptions $option
         $this->assertEquals([['tableId' => 'in.c-API-tests.watched-2']], $updateTrigger['tables']);
     }
 
+    public function testUpdateTriggerParameterCreatedByAdminToken()
+    {
+        $table1 = $this->createTableWithRandomData("watched-1");
+
+        $options = (new TokenCreateOptions())
+            ->addBucketPermission($this->getTestBucketId(), TokenAbstractOptions::BUCKET_PERMISSION_READ);
+
```
[the remark this answers is not in the document]
Comment (fejjo) [reply tests/Common/TriggersTest.php]: Dik upravene

File: tests/Common/TriggersTest.php [hunk 3]
```diff
@@ -298,6 +298,124 @@ public function testUpdateTriggerCreatedByMasterToken(TokenCreateOptions $option
...
+            $clientWithoutAdminTokenWithoutComponentAccess->updateTrigger((int) $trigger['id'], $updateData);
+            self::fail('should fail before');
+        } catch (\Exception $e) {
+            self::assertEquals('Your token does not have sufficient privilege.', $e->getMessage());
+        }
+
+        $options = (new TokenCreateOptions())
+            ->addBucketPermission($this->getTestBucketId(), TokenAbstractOptions::BUCKET_PERMISSION_READ)
+            ->addComponentAccess('keboola.ex-1');
+
```
[the remark this answers is not in the document]
Comment (fejjo) [reply tests/Common/TriggersTest.php]: Dik upravene

File: tests/Common/TriggersTest.php [hunk 4]
```diff
@@ -298,6 +298,124 @@ public function testUpdateTriggerCreatedByMasterToken(TokenCreateOptions $option
...
+            'component' => 'keboola.ex-1',
+            'configurationId' => 543,
+            'coolDownPeriodMinutes' => 15,
+            'runWithTokenId' => $newToken['id'],
+            'tableIds' => [$table1],
+        ];
+
+        $options = (new TokenCreateOptions())
+            ->addBucketPermission($this->getTestBucketId(), TokenAbstractOptions::BUCKET_PERMISSION_READ);
+
```
[the remark this answers is not in the document]
Comment (fejjo) [reply tests/Common/TriggersTest.php]: Dik upravene

File: apiary.apib [hunk 5]
```diff
@@ -8187,7 +8187,7 @@ Gets stats for a given Run Id. Provides the count of imported/exported tables an
 ###Create [POST /v2/storage/triggers/]
 Created a trigger listening on set of tables
 
-Token has to be [*master*](https://help.keboola.com/storage/tokens/) (then `runWithTokenId` can be also master). If it isn't master, it has to have `canManageBuckets` permission or `componentAccess` permission for component `keboola.orchestrator`. Applies also on Update and Delete operation
+Token has to be [*master*](https://help.keboola.com/storage/tokens/) (then `runWithTokenId` can be also master). If it isn't master, it has to have `canManageBuckets` permission or `componentAccess` permission for the component which you want to trigger. This also applies when updating and deleting triggers.
```
Comment (tomasfejfar) [inline apiary.apib:8190]:
```suggestion
Token has to be [*master*](https://help.keboola.com/storage/tokens/) (then `runWithTokenId` can be also master). If it isn't master, it has to have `canManageBuckets` permission or `componentAccess` permission for the component which you want to trigger. This also applies when updating and deleting triggers. When updating the component of trigger you need permissions for both the current component and the component you're updating the trigger to. 
```

**Review by tomasfejfar — COMMENTED** [review]
Ještě mě napadlo, že tohle by tu asi mělo být dospecifikované - je to takový edge case.

File: apiary.apib (hunk 5 above)
Comment (fejjo) [reply apiary.apib:8190]: Jo to je pravda ze to by to bolo matuce. 👍

Comment (fejjo) [reply apiary.apib:8190]: Pridal som

File: tests/Common/TriggersTest.php [hunk 6]
```diff
@@ -265,7 +265,7 @@ public function testUpdateTriggerCreatedByMasterToken(TokenCreateOptions $option
             $clientWithoutAdminTokenWithoutPermissions->updateTrigger((int) $trigger['id'], $updateData);
             self::fail('should fail');
         } catch (Exception $e) {
-            self::assertEquals('Your token does not have sufficient privilege.', $e->getMessage());
+            self::assertEquals('Your token does not have sufficient privilege. Token is not admin token nor does not have canManageBucket permission or does not have access to component(s) "keboola.ex-1, keboola.orchestrator".', $e->getMessage());
```
Comment (tomasfejfar) [inline tests/Common/TriggersTest.php:268]:
```suggestion
            self::assertEquals('Your token does not have sufficient privilege. Token is neither admin token nor does it have canManageBucket permission nor does it have access to component(s) "keboola.ex-1, keboola.orchestrator".', $e->getMessage());
```

File: tests/Common/TriggersTest.php [hunk 7]
```diff
@@ -333,7 +333,7 @@ public function testUpdateTriggerParameterCreatedByAdminToken()
             $clientWithoutAdminTokenWithoutComponentAccess->updateTrigger((int) $trigger['id'], $updateData);
             self::fail('should fail before');
         } catch (\Exception $e) {
-            self::assertEquals('Your token does not have sufficient privilege.', $e->getMessage());
+            self::assertEquals('Your token does not have sufficient privilege. Token is not admin token nor does not have canManageBucket permission or does not have access to component(s) "keboola.ex-1".', $e->getMessage());
```
Comment (tomasfejfar) [inline tests/Common/TriggersTest.php:336]:
```suggestion
            self::assertEquals('Your token does not have sufficient privilege. Token is neither admin token nor does it have canManageBucket permission nor does it have access to component(s) "keboola.ex-1".', $e->getMessage());
```

File: tests/Common/TriggersTest.php [hunk 8]
```diff
@@ -400,16 +400,16 @@ public function testUpdateTriggerComponentWithWrongPermissions(TokenCreateOption
     public function tokenUpdateOptionsProviderInvalid()
     {
         $this->setUp();
-        yield 'Token has only access on actual set component' => [
+        yield 'Token has only access on current set component' => [
             (new TokenCreateOptions())
                 ->addComponentAccess('keboola.orchestrator'),
-            "Your token does not have sufficient privilege.",
+            'Your token does not have sufficient privilege. Token is not admin token nor does not have canManageBucket permission or does not have access to component(s) "keboola.ex-1".',
```
Comment (tomasfejfar) [inline tests/Common/TriggersTest.php:406]:
```suggestion
            'Your token does not have sufficient privilege. Token is neither admin token nor does it have canManageBucket permission nor does it have access to component(s) "keboola.ex-1".',
```

File: tests/Common/TriggersTest.php [hunk 9]
```diff
@@ -400,16 +400,16 @@ public function testUpdateTriggerComponentWithWrongPermissions(TokenCreateOption
...
         ];
 
         yield 'Token has only access on to update set component' => [
             (new TokenCreateOptions())
                 ->addComponentAccess('keboola.ex-1'),
-            "Your token does not have sufficient privilege.",
+            'Your token does not have sufficient privilege. Token is not admin token nor does not have canManageBucket permission or does not have access to component(s) "keboola.orchestrator".',
```
Comment (tomasfejfar) [inline tests/Common/TriggersTest.php:412]:
```suggestion
            'Your token does not have sufficient privilege. Token is neither admin token nor does it have canManageBucket permission nor does it have access to component(s) "keboola.orchestrator".',
```

Comment (tomasfejfar) [reply tests/Common/TriggersTest.php:412]: atd.

**Review by tomasfejfar — APPROVED** [review]
Akorát opravit překlady... to jde asi udělat líp lokálně replacem.

**Comment by fejjo** [comment]
rebase and fix stan
